### ISSUE TITLE
Difference in result of DH_compute_key and EVP_PKEY_derive.

### ISSUE THREAD

**Issue by A-Moeed-B** [openssl/openssl#21315]
Hi,

Our code was doing the DH shared key calculation using the following lines of code pre OpenSSL 3.0
```
 if (!DH_set0_key(priv_key, NULL, new_priv_key) ||
      !DH_set0_pqg(priv_key, new_p, NULL, new_g)) {
    goto err;
  }
  
 int rr = DH_compute_key(s_key, pub_key, priv_key);
  //where new_priv_key, new_priv_key and pub_key are BIGNUMS.
  ```
  The replacement code for deprecating APIs I have written for OpenSSL 3.0  is
  1. Formation of private key.
  ```
EVP_PKEY *form_key_DH(BIGNUM *prime, BIGNUM *generator,BIGNUM *priv_key)
{
	OSSL_PARAM_BLD* param_builder = OSSL_PARAM_BLD_new();
	if (!param_builder) {
		//err
	}
	
	if (!OSSL_PARAM_BLD_push_BN(param_builder, OSSL_PKEY_PARAM_FFC_P, prime)){
		//err
	}
	
	if(!OSSL_PARAM_BLD_push_BN(param_builder, OSSL_PKEY_PARAM_FFC_G, generator)){
		//err
	}
	if(!OSSL_PARAM_BLD_push_BN(param_builder,OSSL_PKEY_PARAM_PRIV_KEY, priv_key)) {
		//err
	}
	
	OSSL_PARAM* params = OSSL_PARAM_BLD_to_param(param_builder);
	if (!params) {
		//err
	}
	
	EVP_PKEY_CTX* dh_key_ctx = EVP_PKEY_CTX_new_from_name(NULL, "DHX", NULL);
	if(!dh_key_ctx){
		//err
	}
	
	if(EVP_PKEY_fromdata_init(dh_key_ctx)<=0){
		//err
	}
	
	EVP_PKEY* dh_key_param=NULL;
	if (EVP_PKEY_fromdata(dh_key_ctx, &dh_key_param,
		EVP_PKEY_KEY_PARAMETERS, params) <= 0) {
		//err
	}
	
	EVP_PKEY_CTX* key_ctx = EVP_PKEY_CTX_new_from_pkey(NULL, dh_key_param, NULL);
	if(!key_ctx) {
		return NULL;
	}
	
	if(EVP_PKEY_keygen_init(key_ctx)<=0) {
		//err
	}
	
	EVP_PKEY* pkey=NULL;
	if(EVP_PKEY_generate(key_ctx,&pkey)<=0) {
		//err
	}
	return pkey;
}
```
Formation of public key: 

```
EVP_PKEY *form_public_key(const BIGNUM* prime, const BIGNUM *generator)
{
	EVP_PKEY* public_key=EVP_PKEY_new();
	OSSL_PARAM_BLD* param_builder = OSSL_PARAM_BLD_new();
	if (!param_builder) {
		//err
	}
	if (!OSSL_PARAM_BLD_push_BN(param_builder, OSSL_PKEY_PARAM_FFC_P, prime)){
		//err
	}
	
	if(!OSSL_PARAM_BLD_push_BN(param_builder, OSSL_PKEY_PARAM_FFC_G, generator)){
		//err
	}
	OSSL_PARAM* params = OSSL_PARAM_BLD_to_param(param_builder);
	if (!params) {
		//err
	}
	EVP_PKEY_CTX* pub_key_ctx = EVP_PKEY_CTX_new_from_name(NULL, "DHX", NULL);
	if(!pub_key_ctx){
		//err
	}
	
	if(EVP_PKEY_fromdata_init(pub_key_ctx)<=0){
		//err
	}
	if (EVP_PKEY_fromdata(pub_key_ctx, &public_key,
		EVP_PKEY_KEY_PARAMETERS, params) <= 0) {
		//err
	}
	return public_key;
}
```
Computation of shared key using EVP_PKEY_derive by sending the priv_key generated by the form_key_DH function.

```
int compute_dh_evp(unsigned char * s_key, const BIGNUM *prime,const BIGNUM *generator, uint8_t *pub, uint16_t pub_len, EVP_PKEY* priv_key)
{
	EVP_PKEY* peerKey = form_public_key(prime,generator);
	EVP_PKEY_set1_encoded_public_key(peerKey, pub, pub_len); //pub and pub_len are used to form the BIGNUM pub_key
	EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(priv_key, NULL);
	if (ctx == NULL) {
		//err
	}
	if(peerKey==NULL)
	{
		//err
	}
	size_t shared_secret_length = 0;
	if (EVP_PKEY_derive_init(ctx) <= 0) {
		//err
	}
	
	
	int res;
	ERR_clear_error();
	if ((res=EVP_PKEY_derive_set_peer(ctx,peerKey )) <= 0) {
		//err
	}
	if (EVP_PKEY_derive(ctx, NULL, &shared_secret_length) <= 0) {
		//err
	}
	//memory to s_key has been assigned before function call.
	if (EVP_PKEY_derive(ctx, s_key, &shared_secret_length) <= 0) {
		//err
	}
	return shared_secret_length;
	
}
```

The functions are successful without any errors. However, the issue is that the result in s_key I'm getting from EVP_PKEY_derive and DH_compute_key is completely different with the same parameters, although the size of the returning data length of the shared_secret key  is same. Can someone point out what I'm doing wrong here?

Thanks!

**Comment by paulidale** [comment]
The EVP variant is going through `DH_compute_key_padded()` not `DH_compute_key()` but that's not going to make a difference.

You are only loading the key parameters (`EVP_PKEY_KEY_PARAMETERS`) when  you load the public and private keys into the PKEY objects.  I suspect they should be `EVP_PKEY_PUBLIC_KEY` and `EVP_PKEY_KEYPAIR` instead.

**Comment by A-Moeed-B** [comment]
I've tried that but still the issue persists.

**Comment by A-Moeed-B** [comment]
> Are you really using the same private key and the same peer key when computing the shared key?
> 
> The code that you've given generates a new private key and does not use the existing priv_key bignum. If you want to use the priv_key, just use EVP_PKEY_fromdata with EVP_PKEY_KEYPAIR.

HI, 
Thanks for pointing the issue. I was able to fix that issue by not generating the key and using the priv_key BIGNUM. However, I'm running into a similar issue while trying to get public key from a generated DH key in a different function. 

Previously I used the code:

```
DH_set0_pqg(dh_key, new_p, NULL, new_g);
DH_set0_key(dh_key, NULL, new_priv_key); 

DH_generate_key(dh_key);
 DH_get0_key(dh_key, &pub_key, NULL);
//able to calculate pub_key 
```
where new_p, new_g,, new_priv_key  and pub_key are BIGNUMS.
Now I use
To generate DH key

```
EVP_PKEY *generate_key_DHX(BIGNUM *prime, BIGNUM *generator,BIGNUM *priv_key)
{
	  OSSL_PARAM_BLD* param_builder = OSSL_PARAM_BLD_new();
	  if (!param_builder) {
		//err
	  }

	  if (!OSSL_PARAM_BLD_push_BN(param_builder, OSSL_PKEY_PARAM_FFC_P, prime)) {
		//err
	  }

	  if(!OSSL_PARAM_BLD_push_BN(param_builder, OSSL_PKEY_PARAM_FFC_G, generator)) {
		//err
	  }
	  if(!OSSL_PARAM_BLD_push_BN(param_builder,OSSL_PKEY_PARAM_PRIV_KEY, priv_key)) {
		//err
	  }

	  OSSL_PARAM* params = OSSL_PARAM_BLD_to_param(param_builder);
	  if (!params) {
		//err
	  }

	  EVP_PKEY_CTX* dh_key_ctx = EVP_PKEY_CTX_new_from_name(NULL, "DHX", NULL);
	  if(!dh_key_ctx){
		//err
	  }

	  if(EVP_PKEY_fromdata_init(dh_key_ctx)<=0){
		//err
	  }

	  EVP_PKEY* dh_key_param=NULL;
	  if (EVP_PKEY_fromdata(dh_key_ctx, &dh_key_param,
		EVP_PKEY_KEYPAIR , params) <= 0) {
		//err
	  }
	  EVP_PKEY_CTX* key_ctx = EVP_PKEY_CTX_new_from_pkey(NULL, dh_key_param, NULL);
	  if(!key_ctx) {
		//err
	  }

	  if(EVP_PKEY_keygen_init(key_ctx)<=0) {
		//err
	  }

	  EVP_PKEY* pkey=NULL;
	  if(EVP_PKEY_generate(key_ctx,&pkey)<=0) {
		//err
	  }
	  return pkey;
}

```

and to get the public key

`EVP_PKEY_get_bn_param(pkey,OSSL_PKEY_PARAM_PUB_KEY,&pub_key)`
OR
`EVP_PKEY_get1_encoded_public_key(pkey, &pub_key_char);`

The result of the original APIs and `EVP_PKEY_get_bn_param` and `EVP_PKEY_get1_encoded_public_key` differ. I have converted the BIGNUM from `EVP_PKEY_get_bn_param` and it matches with `EVP_PKEY_get1_encoded_public_key` but not when I convert the it from  `DH_get0_key`, the resulting buffer is different.

**Comment by A-Moeed-B** [comment]
I'm getting the following error when I don't use `EVP_PKEY_generate` with the pkey and try to get public key using `EVP_PKEY_get1_encoded_public_key()`
`00E71946247F0000:error:02800066:Diffie-Hellman routines:ossl_dh_key2buf:invalid public key:crypto/dh/dh_key.c:414`

**Comment by t8m** [comment]
Hmm... unfortunately the public key is not computed at import time automatically and we do not really have a way to compute it.

**Comment by A-Moeed-B** [comment]
If that is the case then what is the replacement for this set of  older APIs? I'm getting the pub_key using the generated priv_key using the `EVP_PKEY_fromdata` and `EVP_PKEY_generate` but the resulting buffer is different from `DH_get0_key()` an `DH_generate.`
```
DH_set0_pqg(dh_key, new_p, NULL, new_g);
DH_set0_key(dh_key, NULL, new_priv_key); 

DH_generate_key(dh_key);
DH_get0_key(dh_key, &pub_key, NULL);
```

**Comment by t8m** [comment]
There is currently no replacement for this functionality, which is a bug.

There was a vote passed https://github.com/openssl/technical-policies/issues/54 that we should do this automatically on import, but it was not yet implemented.

**Comment by KhepriRa** [comment]
Hi, 
I have a similar issue and I'd like to know if it has been implemented in 3.5 ?

**Comment by huwcbjones** [comment]
I too am the next unfortunate soul trying _not_ use the deprecated APIs, but finding that the new APIs aren't actually _feature complete_  - which I note is different to a bug (unintended behaviour), whereas this is a complete lack of functionality.

The answer is to bounce through PKCS8 (as repulsive as some might find) as the conversion process causes the public key to be recomputed!

TL;DR
```c
EVP_PKEY* dh_key_param=NULL;
if (
	EVP_PKEY_fromdata(
		dh_key_ctx, &dh_key_param, EVP_PKEY_KEYPAIR, params
	) <= 0
) {
	//err
}
PKCS8_PRIV_KEY_INFO *pkcs8 = EVP_PKEY2PKCS8(dh_key_param);
EVP_PKEY_free(dh_key_param)
dh_key_param = EVP_PKCS82PKEY(pkcs8);
PKCS8_PRIV_KEY_INFO_free(pkcs8)
```
